sklearn/datasets/tests/test_openml.py

```diff
@@ -335,39 +160,216 @@ def _mock_urlopen(request, *args, **kwargs):
+        # cpu
+        (561, {"data_id": 561}, 209, 7, 1),
+        (561, {"name": "cpu", "version": 1}, 209, 7, 1),
+        # emotions
+        (40589, {"data_id": 40589}, 13, 72, 6),
+        # adult-census
+        (1119, {"data_id": 1119}, 10, 14, 1),
+        (1119, {"name": "adult-census"}, 10, 14, 1),
+        # miceprotein
+        (40966, {"data_id": 40966}, 7, 77, 1),
+        (40966, {"name": "MiceProtein"}, 7, 77, 1),
+    ],
 )
-def test_feature_to_dtype_error(feature):
-    msg = "Unsupported feature: {}".format(feature)
-    with pytest.raises(ValueError, match=msg):
-        _feature_to_dtype(feature)
+@pytest.mark.parametrize("parser", ["liac-arff", "pandas"])
+def test_fetch_openml_as_frame_false(
+    monkeypatch,
+    data_id,
+    dataset_params,
+    n_samples,
+    n_features,
+    n_targets,
+    parser,
+):
+    """Check the behaviour of `fetch_openml` with `as_frame=False`.
+
+    Fetch by ID and/or name (depending if the file was previously cached).
+    """
+    pytest.importorskip("pandas")
+
+    _monkey_patch_webbased_functions(monkeypatch, data_id, gzip_response=True)
+    bunch = fetch_openml(
+        as_frame=False,
+        cache=False,
+        parser=parser,
+        **dataset_params,
+    )
+    assert int(bunch.details["id"]) == data_id
+    assert isinstance(bunch, Bunch)
+
+    assert bunch.frame is None
+
+    assert isinstance(bunch.data, np.ndarray)
+    assert bunch.data.shape == (n_samples, n_features)
+
+    assert isinstance(bunch.target, np.ndarray)
+    if n_targets == 1:
+        assert bunch.target.shape == (n_samples,)
+    else:
+        assert bunch.target.shape == (n_samples, n_targets)
+
+    assert isinstance(bunch.categories, dict)
+
+
+# Known failure of PyPy for OpenML. See the following issue:
+# https://github.com/scikit-learn/scikit-learn/issues/18906
+@fails_if_pypy
+@pytest.mark.parametrize("data_id", [61, 1119])
+def test_fetch_openml_consistency_parser(monkeypatch, data_id):
+    """Check the consistency of the LIAC-ARFF and pandas parser.
+
+    In the future, we should test for titanic dataset. However, LIAC-ARFF
+    always use string dtype for categorical features whereas pandas infer the
+    data type and thus can use integer or float dtype. Fixing this bug in
+    LIAC-ARFF would allow to check the consistency in the future.
+    """
+    pd = pytest.importorskip("pandas")
+
+    _monkey_patch_webbased_functions(monkeypatch, data_id, gzip_response=True)
+    bunch_liac = fetch_openml(
+        data_id=data_id,
+        as_frame=True,
+        cache=False,
+        parser="liac-arff",
+    )
+    bunch_pandas = fetch_openml(
+        data_id=data_id,
+        as_frame=True,
+        cache=False,
+        parser="pandas",
+    )
+
+    frame_liac, frame_pandas = bunch_liac.frame, bunch_pandas.frame
+    frame_pandas_casted = frame_pandas.astype(frame_liac.dtypes)
+
+    pd.testing.assert_frame_equal(frame_liac, frame_pandas_casted)
```

Comment:
We can test the tiantic case if we rename the categories from the liac parser here:

```python
    frame_liac, frame_pandas = bunch_liac.frame, bunch_pandas.frame

    # convert dtype of liac parser to make it comparable to the pandas parse
    def convert_dtype(series):
        pandas_series = frame_pandas[series.name]
        if pd.api.types.is_numeric_dtype(pandas_series):
            return series.astype(pandas_series.dtype)
        elif pd.api.types.is_categorical_dtype(pandas_series):
            # Compare categorical features by converting categorical
            # liac uses strings to denote the categories, we rename the categories
            # to make them comparable to the pandas parser
            return series.cat.rename_categories(pandas_series.cat.categories)
        else:
            return series

    frame_liac = frame_liac.apply(convert_dtype)
    pd.testing.assert_frame_equal(frame_liac, frame_pandas)
```

As long as the order of `categories` is the same between the two parsers, I think this should work. The snippet above could be a little too hacky tho.